File: ingest-service/src/main/java/dev/xj16/beacon/ingest/ElasticsearchConfig.java
package dev.xj16.beacon.ingest;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.json.jackson.JacksonJsonpMapper;
import co.elastic.clients.transport.ElasticsearchTransport;
import co.elastic.clients.transport.rest_client.RestClientTransport;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import org.apache.http.HttpHost;
import org.elasticsearch.client.RestClient;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.net.URI;

/**
 * Builds the Elasticsearch Java client from the configured URI. The JSONP mapper reuses a Jackson
 * {@link ObjectMapper} that understands {@code java.time} types so {@code Instant}s serialize as
 * ISO-8601 dates rather than epoch arrays.
 */
@Configuration
public class ElasticsearchConfig {

    @Bean(destroyMethod = "close")
    @ConditionalOnMissingBean
    public RestClient elasticsearchRestClient(
            @Value("${beacon.elasticsearch.uri:http://localhost:9200}") String uri) {
        URI parsed = URI.create(uri);
        int port = parsed.getPort() == -1 ? 9200 : parsed.getPort();
        String scheme = parsed.getScheme() == null ? "http" : parsed.getScheme();
        String host = parsed.getHost() == null ? "localhost" : parsed.getHost();
        return RestClient.builder(new HttpHost(host, port, scheme)).build();
    }

    @Bean
    @ConditionalOnMissingBean
    public ElasticsearchClient elasticsearchClient(RestClient restClient) {
        // Serialize Instant as ISO-8601 strings (not epoch floats) so the Elasticsearch `date`
        // mapping accepts them; a float timestamp triggers a mapper_parsing_exception on index.
        ObjectMapper mapper = new ObjectMapper()
                .registerModule(new JavaTimeModule())
                .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
        ElasticsearchTransport transport =
                new RestClientTransport(restClient, new JacksonJsonpMapper(mapper));
        return new ElasticsearchClient(transport);
    }
}

File: query-api/src/main/kotlin/dev/xj16/beacon/query/ElasticsearchConfig.kt
package dev.xj16.beacon.query

import co.elastic.clients.elasticsearch.ElasticsearchClient
import co.elastic.clients.json.jackson.JacksonJsonpMapper
import co.elastic.clients.transport.rest_client.RestClientTransport
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.KotlinModule
import org.apache.http.HttpHost
import org.elasticsearch.client.RestClient
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.net.URI

/** Wires the Elasticsearch Java client from the configured URI. */
@Configuration
class ElasticsearchConfig {

    @Bean(destroyMethod = "close")
    fun restClient(@Value("\${beacon.elasticsearch.uri:http://localhost:9200}") uri: String): RestClient {
        val parsed = URI.create(uri)
        val host = parsed.host ?: "localhost"
        val port = if (parsed.port == -1) 9200 else parsed.port
        val scheme = parsed.scheme ?: "http"
        return RestClient.builder(HttpHost(host, port, scheme)).build()
    }

    @Bean
    fun elasticsearchClient(restClient: RestClient): ElasticsearchClient {
        val mapper = ObjectMapper()
            .registerModule(JavaTimeModule())
            .registerModule(KotlinModule.Builder().build())
        val transport = RestClientTransport(restClient, JacksonJsonpMapper(mapper))
        return ElasticsearchClient(transport)
    }
}
